File: leetcode/scala/3.longest-substring-without-repeating-characters.scala
import scala.collection.mutable

object Solution {
  def lengthOfLongestSubstring(s: String): Int = {
    var begin = -1
    var res = 0
    val map = new mutable.HashMap[Char, Int]()
    for (i <- Range(0, s.length)) {
      // 已经出现且在现在的这个字串里
      if (map.contains(s(i)) && map(s(i)) > begin) {
        begin = map(s(i))
      }

      res = Math.max(res, i - begin)
      map(s(i)) = i
    }
    res

  }
}

//  0 作为下标的索引
import scala.collection.mutable

object Solution {
  def lengthOfLongestSubstring(s: String): Int = {
    var begin = 0
    var res = 0
    val map = new mutable.HashMap[Char, Int]()
    for (i <- Range(0, s.length)) {
      if (map.contains(s(i)) && map(s(i)) >= begin) {
        begin = map(s(i)) + 1
      }

      res = Math.max(res, i - begin + 1)
      map(s(i)) = i
    }
    res

  }
}

File: leetcode/java/3.longest-substring-without-repeating-characters.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        int sub_start = 0;
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            if (map.containsKey(s.charAt(i)) && map.get(s.charAt(i)) >= sub_start) {
                sub_start = map.get(s.charAt(i)) + 1;
            }
            map.put(s.charAt(i), i);
            res = Math.max(res, i - sub_start + 1);
        }

        return res;

    }
}
